azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -44,10 +51,79 @@ def __init__(self, callback):
                 "Cannot instantiate PipelineOperation object.  You need to use a derived class"
             )
         self.name = self.__class__.__name__
-        self.callback = callback
+        self.callbacks = []
         self.needs_connection = False
         self.completed = False
 
+        self.callbacks.append(callback)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def add_callback(self, callback):
+        self.callbacks.append(callback)
+
+    @pipeline_thread.runs_on_pipeline_thread
+    def complete(self, error=None):
+        if error:
+            logger.error("{}: completing with error {}".format(self.name, error))
+        else:
+            logger.debug("{}: completing without error".format(self.name))
+
+        if self.completed:
+            logger.error("{}: has already been completed!".format(self.name))
+            e = pipeline_exceptions.OperationError(
+                "Attempting to complete an already-completed operation: {}".format(self.name)
+            )
+            # This could happen in a foreground or background thread, so err on the side of caution
+            # and send it to the background handler.
+            handle_exceptions.handle_background_exception(e)
+        else:
+            self.completed = True
+
+            while self.callbacks:
+                if not self.completed:
+                    logger.debug(
+                        "{}: Operation no longer complete! Suspending completion".format(self.name)
+                    )
+                    break
+                callback = self.callbacks.pop()
+                try:
+                    callback(op=self, error=error)
```

Comment:
>error=error [](start = 38, length = 11)

I like this code, but there are 2 big implications here:

1. You suspend callback processing by "uncompleting" the op, which also clears the error.
2. Inside a callback, you can't change the error to a different error, and you also can't change success to error.

I think these are both perfectly acceptable, and maybe even desirable.  I just want to put it out there and keep it in mind.

Reply:
Added some further documentation, I think this is clearer now.

---
In reply to: [345991747](https://github.com/Azure/azure-iot-sdk-python/pull/392#discussion_r345991747) [](ancestors = 345991747)